### PR TITLE
Reduce tensor-parallel logits communication with a rank-0 gather

### LINKED ISSUE #3365: [Feature] support `gather` instead of `all_gather` when gathering the logits
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

We noticed that in the `_get_logits` function of vllm, `gather` instead of `all_gather` will be used under certain conditions (the main condition is that for non-tpu devices):
Code link:

- [logits = tensor_model_parallel_gather(logits)](https://github.com/vllm-project/vllm/blob/6e1fc61f0fb90c37f0d4a1a8f76235a6e4e1103c/vllm/model_executor/layers/logits_processor.py#L101C22-L101C50)

- [condition of whether using `all_gather` or `gather`](https://github.com/vllm-project/vllm/blob/6e1fc61f0fb90c37f0d4a1a8f76235a6e4e1103c/vllm/model_executor/layers/logits_processor.py#L53-L57)

The change from using `all_gather` to `gather` is initially added in this PR for your reference: https://github.com/vllm-project/vllm/pull/2221.

While in SGLang, we see currently `all_gather` is always used:
https://github.com/sgl-project/sglang/blob/e868d0b60eb2d435c5599165f787bca06bdc9c3d/python/sglang/srt/layers/logits_processor.py#L246

Does SGLang have the plan to add `gather` instead of only `all_gather` when gathering the logits? Per the practice in vllm, using `gather` seems to have better performance than `all_gather` on devices which have `gather` support.

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by SwayamInSync**
Closes #3365.

`LogitsProcessor` currently `all_gather`s the vocab-parallel logits, so every TP rank materializes the full `[num_tokens, vocab]` tensor and redundantly runs the full-vocab sampling. For large-vocab models at higher TP, this logits gather is a real chunk of the decode step.

We can instead `gather` the logits onto TP rank 0 only: rank 0 samples, and the tiny sampled token ids are broadcast back to the other ranks. This roughly halves the logits communication and removes the redundant per-rank sampling compute. 

This is closely related to our internal work on reducing decode-time communication, and lands it as an opt-in path in the open-source runtime. Enabled via `--enable-gather-logits`; plain tensor-parallel path only (ignored under DP-attention, `--enable-dp-lm-head`, and speculative decoding).

## Modifications

The gather itself needed to stay capturable inside the decode CUDA graph, so `GroupCoordinator.gather` now uses grouped pynccl `send`/`recv` on the device path (`torch.distributed.gather` can't be captured). On top of that, `LogitsProcessor`
gathers onto rank 0 when the flag is set -- the other ranks simply get `None` back and skip the rest of the logits post-processing. Sampling moves accordingly: `ModelRunner.sample` samples on rank 0 and broadcasts the token ids to the TP group, and `Sampler` gets a `skip_tp_sync` flag so its usual cross-TP token all-reduce doesn't run in this path. Finally there's the new `--enable-gather-logits` server arg, which is turned off automatically when
speculative decoding is active.

## Accuracy Tests

- Greedy output parity vs the all-gather baseline (byte-for-byte identical tokens).
- GSM8K + JSON/Regex constrained decoding, CUDA graph enabled and eager.
- Multi-GPU collective test: gather == all_gather on the destination rank
  (eager + CUDA graph, fp16/bf16/fp32); `None` on the other ranks.
  
 ## Speed Tests and Profiling

**Hardware:** 4×A100 80GB PCIe ·

The benefit of `gather` over `all_gather` grows with TP degree and decode batch size, where logits communication is a larger share of each step. At **TP=4 under high concurrency** the official serving benchmark shows a consistent ~6% end-to-end gain:

**Qwen2.5-3B-Instruct · TP=4 · concurrency 256 · 1024 prompts · input 512 / output 1024**

| Metric | Baseline (`all_gather`) | Gather | Δ |
|---|---:|---:|---:|
| Request throughput (req/s) | 17.67 | 18.77 | **+6.2%** |
| Total token throughput (tok/s) | 13,574 | 14,425 | **+6.3%** |
| Mean TPOT (ms) | 26.56 | 24.90 | **−6.3%** |
| Median E2E latency (ms) | 12,739 | 11,878 | **−6.8%** |

Reproduce:

```bash
python -m sglang.launch_server --model-path Qwen/Qwen2.5-3B-Instruct --tp 4 --port 21000 [--enable-gather-logits]

python -m sglang.bench_serving --backend sglang --host 127.0.0.1 --port 21000 \
  --dataset-name random --random-input-len 512 --random-output-len 1024 \
  --num-prompts 1024 --max-concurrency 256
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29390057910](https://github.com/sgl-project/sglang/actions/runs/29390057910)
Latest PR Test (Extra): :x: [Run #29390057821](https://github.com/sgl-project/sglang/actions/runs/29390057821)

### CODE REVIEW

**Comment by SwayamInSync** [comment]
cc: @zhaochenyang20

**Review by kpham-sgl — COMMENTED** [review]
Hi @SwayamInSync, vocab parallel all-gather now uses `triton_symm_mem_ag` (for symm-mem available hw) which is a much stronger baseline than NCCL all-gather. Do you think `gather` is still needed?

**Comment by SwayamInSync** [comment]
> Hi @SwayamInSync, vocab parallel all-gather now uses `triton_symm_mem_ag` (for symm-mem available hw) which is a much stronger baseline than NCCL all-gather. Do you think `gather` is still needed?

Thanks @kpham-sgl, that's a fair push, I re-benchmarked on 8x B200 (NVLink5/NVSwitch), and a plain NCCL gather does **not** beat it; end-to-end the logits gather is only about 3% of the decode step

Two reasons I think it's still worth keeping, scoped as a **fallback that's never a regression**:

**1. Non-symm-mem / PCIe HW** (no multicast): the gather still beats all-gather. This is the original A100-PCIe result in this PR (+6%).

**2. On symm-mem HW we can match/beat the baseline** with a symm-mem *scatter-to-rank-0* instead of pynccl. 
One-line idea: it's your `_all_gather_kernel_inner` with `multimem.st` -> `st` and `multicast_ptr` -> `buffer_ptrs[0]`, so each rank writes its shard straight into rank 0's symmetric buffer at its column offset, lands in `[T,V]` with no reshape, fully CUDA-graph-capturable, no NCCL. I prototyped it; bit-exact vs all-gather (eager + graph).

Per-op collective latency, CUDA graph, TP=8, `[256 tokens x 152k vocab]` bf16 (lower = better):

| collective | latency (ms) |
|---|---|
| NCCL all-gather (PCIe-class baseline) | 0.244 |
| **triton symm-mem all-gather (your baseline)** | **0.119** |
| pynccl gather (this PR, fallback path) | 0.183 |
| **symm-mem scatter gather (proposed)** | **0.105** |

The scatter's edge over the multimem AG is stable at ~1.14x across decode batch 128–4096, and it grows with TP (roughly tie at TP=2, ~1.14x at TP=8)

So on symm-mem HW the scatter gather (0.105) is on par with or a bit faster than the multimem AG (0.119); on no-multicast HW the gather (0.183) beats all-gather (0.244). Proposed shape mirrors `MultimemAllGatherer`'s own NCCL fallback:

```
symm-mem + multicast -> symm-mem scatter-to-rank-0
else                 -> pynccl gather
```

Given the e2e delta on symm-mem HW is small, I'm happy to scope this however you prefer: land it opt-in with the scatter path so it's parity-or-better everywhere, or keep it PCIe-only. Would the scatter-kernel version change your view, and do you want me to push it into this PR?
